crates/control/src/localization.rs

```diff
@@ -208,26 +208,44 @@ impl Localization {
                     None => {}
                 };
             }
-            (PrimaryState::Penalized, _, _)
-                if primary_state != PrimaryState::Penalized
-                    && (self.time_when_penalized_clicked.map_or(true, |time| {
+            (PrimaryState::Penalized, _, _) if primary_state != PrimaryState::Penalized => {
+                if self.was_picked_up_while_penalized
+                    || self.time_when_penalized_clicked.map_or(true, |time| {
                         context
                             .cycle_time
                             .start_time
                             .duration_since(time)
                             .expect("time ran backwards")
                             > *context.tentative_penalized_duration
-                    }) || !context.has_ground_contact) =>
-            {
-                if self.is_penalized_with_motion_in_set_or_initial {
-                    if self.was_picked_up_while_penalized_with_motion_in_set_or_initial {
-                        self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
-
+                    })
+                {
+                    if self.is_penalized_with_motion_in_set_or_initial {
+                        if self.was_picked_up_while_penalized {
+                            self.hypotheses = take(&mut self.hypotheses_when_entered_playing);
+
+                            let penalized_poses = generate_penalized_poses(
+                                context.field_dimensions,
+                                *context.penalized_distance,
+                            );
+                            self.hypotheses_when_entered_playing = penalized_poses
+                                .into_iter()
+                                .map(|pose| {
+                                    ScoredPose::from_isometry(
+                                        pose,
+                                        *context.penalized_hypothesis_covariance,
+                                        *context.initial_hypothesis_score,
+                                    )
+                                })
+                                .collect();
+                        }
```

Comment:
Why do you regenerate this and refill this?  Would it be sufficient to copy the `hypotheses_when_entered_playing`?

Reply:
Good catch, done.